bootstrap/021_external_functions.sql

```diff
@@ -104,19 +104,71 @@ create function if not exists internal.get_ef_url()
     as
     'throw "You must configure a Sundeck token to use this.";';
 
+create function if not exists internal.get_tenant_url()
+    returns string
+    language javascript
+    as
+    'throw "You must configure a Sundeck token to use this.";';
+
 create function if not exists internal.get_ef_token()
     returns string
     language javascript
     as
     'throw "You must configure a Sundeck token to use this.";';
 
+create or replace procedure internal.setup_ef_url(url string) RETURNS STRING LANGUAGE SQL AS
+BEGIN
+    execute immediate 'create or replace function internal.get_ef_url() returns string as \'\\\'' || url || '\\\'\';';
+END;
+
 
 create or replace procedure internal.setup_sundeck_token(url string, token string) RETURNS STRING LANGUAGE SQL AS
 BEGIN
     execute immediate 'create or replace function internal.get_ef_url() returns string as \'\\\'' || url || '\\\'\';';
     execute immediate 'create or replace function internal.get_ef_token() returns string as \'\\\'' || token || '\\\'\';';
 END;
 
+create function if not exists internal.ef_register_tenant(request object)
+    returns object
+    language javascript
+    as
+    'throw "tenant register requires api gateway to be configured";';
+
+
+create or replace function internal.wrapper_register_tenant(request object)
+    returns object
+    immutable
+as
+$$
+    iff(length(internal.ef_register_tenant(request):error) != 0,
+        internal.throw_exception(internal.ef_register_tenant(request):error),
+        internal.ef_register_tenant(request))::object
+$$;
+
+create or replace function tools.register_tenant(client_id varchar, client_secret varchar)
```

Comment:
I believe that the `TOOLS` schema is exposed to users with functions that users can expect to use as OpsCenter users (e.g. tools that help the users understand and user snowflake better). This seems more like an implementation detail of setting up the Sundeck account. Could we move this into `INTERNAL`?

IMO we should strive for a no-arg procedure in `ADMIN` that would create the necessary objects and then call the external function. If the account was already set up, it can return whatever is necessary; otherwise, it's created for the user.